File: src/main/java/org/algohub/engine/judge/JudgeTask.java
package org.algohub.engine.judge;

import org.algohub.engine.type.InternalTestCase;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.Callable;

class JudgeTask implements Callable<List<Boolean>> {

    private static final Logger LOG = LoggerFactory.getLogger(JudgeTask.class);

    private final Object clazz;
    private final Method method;
    private final InternalTestCase[] testCases;

    JudgeTask(Object clazz, Method method, InternalTestCase[] testCases) {
        this.clazz = clazz;
        this.method = method;
        this.testCases = testCases;
    }

    @Override
    public List<Boolean> call() throws Exception {
        return run();
    }

    List<Boolean> run() throws InterruptedException {
        ArrayList<Boolean> results = new ArrayList<>();

        for (final InternalTestCase internalTestCase : testCases) {
            final boolean isCorrect = judge(clazz, method, internalTestCase);

            results.add(isCorrect);
        }

        return results;
    }

    private static boolean judge(final Object clazz, final Method method,
                                 final InternalTestCase testCase) throws InterruptedException {
        final Object output;
        Object[] input = null;
        try {
            input = testCase.getInput();
            output = method.invoke(clazz, input);
        } catch (IllegalAccessException | InvocationTargetException e) {
            Throwable cause = getCause(e);
            LOG.error("Error during processing of solution", cause);
            throw new InterruptedException(cause.getClass().getName() + ": " + cause.getMessage());
        } catch (IllegalArgumentException e) {
            if (input != null) {
                LOG.error("Input count: " + input.length);
                Arrays.asList(input).forEach(
                        item -> LOG.error("Input type: " + item.getClass().toGenericString())
                );
            }

            LOG.error("Method parameters count: " + method.getParameters().length);
            Arrays.asList(method.getParameters()).forEach(
                    item -> LOG.error("Parameter type: " + item.getClass().toGenericString())
            );

            Throwable cause = getCause(e);
            LOG.error("Error during processing of solution", cause);

            throw new InterruptedException(cause.getClass().getName() + ": " + cause.getMessage());
        }

        if (testCase.returnsVoid()) {
            return BetterObjects.equalForObjectsOrArrays(testCase.getOutput(), input[0]);
        }

        return BetterObjects.equalForObjectsOrArrays(testCase.getOutput(), output);
    }

    private static Throwable getCause(Throwable e) {
        Throwable cause;
        Throwable result = e;

        while(null != (cause = result.getCause())  && (result != cause) ) {
            result = cause;
        }
        return result;
    }
}

File: src/main/kotlin/org/algohub/engine/judge/JudgeTask.kt
package org.algohub.engine.judge

import org.slf4j.LoggerFactory
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.concurrent.Callable

internal class JudgeTask(private val clazz: Any, private val method: Method, private val testCases: Array<InternalTestCase>) : Callable<List<Boolean>> {

    private val LOG = LoggerFactory.getLogger(JudgeTask::class.java)

    @Throws(Exception::class)
    override fun call(): List<Boolean> = testCases.map { judge(clazz, method, it) }

    @Throws(InterruptedException::class)
    private fun judge(clazz: Any, method: Method, testCase: InternalTestCase): Boolean {

        val output: Any?
        val input: Array<Any?>

        try {
            input = testCase.input
            output = method.invoke(clazz, *input)
        } catch (e: IllegalAccessException) {
            val cause = getCause(e)
            LOG.error("Error during processing of solution", cause)
            throw InterruptedException(cause.javaClass.name + ": " + cause.message)
        } catch (e: InvocationTargetException) {
            val cause = getCause(e)

            if (cause is NullPointerException) {
                return false
            }

            LOG.error("Error during processing of solution", cause)
            throw InterruptedException(cause.javaClass.name + ": " + cause.message)
        } catch (e: IllegalArgumentException) {
            val cause = getCause(e)
            LOG.error("Error during processing of solution", cause)
            throw InterruptedException(cause.javaClass.name + ": " + cause.message)
        }

        return BetterObjects.equalForObjectsOrArrays(
                testCase.output,
                if (testCase.returnsVoid) input[0] else output
        )
    }

    private fun getCause(e: Throwable): Throwable {
        var cause: Throwable? = e.cause
        var result = e

        while (null != cause && result !== cause) {
            result = cause
            cause = result.cause
        }
        return result
    }
}
